### ISSUE TITLE
Safe TLD configurable

### ISSUE THREAD

**Issue by Moutard3** [cakephp/debug_kit#617]
The array containing the safe TLDs should be configurable with something like this:
`Configure::write('DebugKit', ['safeTLD' => ["localhost", "dev", "test", "example", "local","somethingelse"]]);`
I think that's a useful feature tu add one more layer to disable DebugKit in prod in case debug mode is not disabled, but it can't be used if not using a "standard" dev TLD (because we need to forceEnable it).
In my case I'm using `.loc` and Laragon is using `.oo` by default.

**Comment by markstory** [comment]
If this config value is empty, would the code default to the current logic?

**Comment by dereuromark** [comment]
It would be nice if that would disable it, solving the issue for people that dont like to use this overprotection. As it really can cause more harm than good for most setups.

**Comment by dereuromark** [comment]
I explained that already in the other ticket.
It creates for those people (that are not experienced enough) the false assumption that all is good, but in fact debug mode is silently still on (without them knowing), and then every stack trace can reveal far more uncensored information about the underlying system.

**Comment by markstory** [comment]
> In this case it is 10x more helpful to display it, and making sure that person knows that turning debug mode off is the right call (and fixes the root issue), not hiding the toolbar (fixing the symtom only).

I don't disagree with you here. The issue we were trying to solve is CakePHP (the project) being blamed/shamed when our users deploy applications in debug mode. It is *us* and not the individual developer who is blamed for these kinds of mistakes.

**Comment by dereuromark** [comment]
Being a bit over-sensitive over a theoretical blame game is one thing, but actually compromising security of many apps even more this way due to the false security is a total different one.
For me this was a clear security issue opened - instead of closed. Because before it was a "baah, that was really obvious you fucked it up", now it is "CakePHP is dangerously hiding the toolbar by magic, that caused the accidental (unwilling) debug mode to expose lots of sensitive info over a long timeframe".

To sum things up:
For the first the user (beginner?) is to blame (not the framework clearly)
For the 2nd it gets now really hard to tell beginners that it was obvious, when its not (now the actual blame/shame is on the framework for making it that dangerous).

I stated those before the merge. And I was ignored :/

**Comment by dereuromark** [comment]
To actually give a GOOD example of what should have been done:

On production (isSuspiciallyProduction true) the Icon should be even fatter, blinking, a huge arrow pointing to it with the bubble text of

> HELLO, you seem to be productive, but you still have debug on on"
For security reasons you MUST deactive it now.

That will make it even more super-clear what the issue is, and everyone is happy.

**Comment by markstory** [comment]
> On production (isSuspiciallyProduction true) the Icon should be even fatter, blinking, a huge arrow pointing to it with the bubble text of

This is an interesting line of thought. What if we didn't let the application render if debugkit was enabled on a non-whitelisted domain?

**Comment by jeremyharris** [comment]
A big red full-page error like when you're visiting a site with an invalid SSL cert or malware site would be an option.

**Comment by lorenzo** [comment]
On the topic of what this feature is for, I’ll explain once again. Beyond the idea that we don’t want shame as a project for mistakes careless developers do, this is protecting people like myself who have in the past deployed a visible debugkit to production by mistake, if only for a few minutes.

The risk of having debug on is only made dangerous when hitting certain error pages, or when using certain debug features, but that is not what the general public will see. On the other hand, debugkit is visible to everyone, which makes it more dangerous to discover.

An error page will only leak selected parts of your app, and maybe one faulty or malformed query; but debugkit will show the full structure of you application folder, a big chunk of your database schema, and the technologies you use behind scenes.



I think the red page of death idea is worth exploring, but only when using debug aware features, such as de debug functions, and the helpful exception pages.

In the meantime, @Moutard3 I offer you a couple solutions that we could make official recommendations if you think they solve your issue:

* Align with a bigger group of developers using those TLDs for development. I don’t think that is going to change your workflow too much.
* export a environment variable ‘SHOW_DEBUGKIT’ in a place that is visible to your local web server. Then pass that variable to the forceEnable flag I’m debugkit.

The latter option will always be safe as long as you don’t add the same env var with value 1 production.

I’ll let others vote for your suggestion of adding more sage TLDs, but personally I think the force enable flag is enough to cover your use case.

**Comment by markstory** [comment]
@lorenzo The death page could be triggered from the middleware if the host looks unsafe, rendering the application unusable. This makes the mistake painfully obvious to the developer as their app will be 'down'.

To me this helps solve the original problem of making accidental mistakes easier to catch and protects application secrets/credentials as the application is effectively locked down.

**Comment by jorisvaesen** [comment]
I'm also in favor of making the app inaccessible when debug is on in production. But shouldn't this be build into ErrorHandlerMiddleware since you can deploy a CakePHP app without debugkit. I'm afraid inexperienced developers would just remove the Plugin::load('DebugKit') line in bootstrap.php to fix it for them. Putting it into ErrorHandlerMiddleware (which is enabled by default in Application.php) seems like a safer option...

**Comment by davidyell** [comment]
References #616 #611 #608 #622 #613 #595 and I'm sure there are a few more 😄

**Comment by highstrike** [comment]
### I feel that all these solutions are now created for problems that shouldn't exist in the first place

The **developer** should have the choice to either deploy with debug on or off on production. 
Just like the **developer** has a choice to upload the files on production with the composer `--no-dev` flag or without.

IMHO the original feature #595 just adds more problems than it tries to solve, and because if it now we have things like:
- silently running with debug on without the possibility of clearly seeing it now - @dereuromark 
- safe TLDs not covering all the use cases - @davidyell 
- make debugkit icon blink, huge arrows, bubble of texts - @dereuromark 
- having the `forceEnable` and now the `safeTld` extra options in order to fix what #595 added by *default* - @lorenzo @dereuromark 
- preventing the site from rendering at all because of the debug flag + aditional logic in the `ErrorHandlerMiddleware` (this is not just a debugkit *plugin* change now, its a cakephp *framework* one) - @jeremyharris @jorisvaesen 

## More
Why would my application break with a red error page just because I wanted to have debug enabled on production for 5 minutes to record some logs that i wouldn't log otherwise? You guys will probably tell me, well, no, make another flag for that. Why would I as a developer have to change my existing application because of a new feature being added... *to a PLUGIN* that I don't even install on Prod.

## TL;DR
DebugKit's mission shouldn't be that to protect developers from themselves. It's a plugin that does debugging, if you forget it on production that's on you as a developer, not on the plugin or on the framework.

**Comment by markstory** [comment]
> Why would I as a developer have to change my existing application because of a new feature being added... to a PLUGIN that I don't even install on Prod.

If *you* are not installing this plugin on production you won't have to make any changes.

I personally don't think we should be changing the cakephp exception handling for this case. If someone wants to deploy with debug on that's their own choice. I do feel that debugkit should do its best to protect its users though. Because the plugin can expose sensitive data by accident it should also do its best to keep data safe.

The error page that I proposed would be a debugkit only change.

**Comment by mehov** [comment]
I couldn't get it to work [using `safeTld`](https://github.com/cakephp/debug_kit/issues/617) with either of these:

    Configure::write('DebugKit.safeTLD', ['dev.example.com']); // didn't help
    Configure::write('DebugKit', ['safeTLD' => ['dev.example.com']]); // didn't help either
    Plugin::load('DebugKit', ['bootstrap' => true]);

However, the [good old `forceEnable`](https://book.cakephp.org/3.0/en/debug-kit.html#configuration) worked:

    Configure::write('DebugKit.forceEnable', true);
    Plugin::load('DebugKit', ['bootstrap' => true]);

**Comment by dereuromark** [comment]
mind your casing saveTld not saveTLD.
